### DISCUSSION TITLE
Script to add hotkey support to apps lacking it

### DISCUSSION THREAD

**Discussion by lcrespom** [johnlindquist/kit#285] [Q&A] [answered]
Hello,

I am using an old e-mail app that has been (poorly) ported to Mac and lacks proper hotkey support for basic functions such as mail send, mark as read/unread, etc. This means having to navigate through menus with the mouse, which is much slower.

I would like to use Script Kit to:
- Detect a certain combination of keys at any time, in the background (e.g. Ctrl+Shift+PageUp)
- Check for the current application and focused window title
- If the app and window title match the expected values (e.g. the e-mail app is focused and its window title contains "Inbox"), then
- Select a specific submenu item of a specific menu (e.g. Actions => Mark as => Mark selected as unread). Or alternatively, simulate the user pushing a sequence of keystrokes to navigate through the menu (e.g. Ctrl+F2, "c", Right, Down, "m", Down, Down, Right, Return).

I am proficient in JavaScript but totally new to ScriptKit. If you could give me the list of ScriptKit APIs to do all the above steps, I can write the necessary logic around it.

Thank you very much :-)

**Comment by johnlindquist** [comment, marked as answer]
Here you go, not Script Kit, but it's how I would do it 😇

https://user-images.githubusercontent.com/36073/118866482-457c0a80-b89f-11eb-8415-295c67160c50.mp4

**Reply by lcrespom** [reply]
Impressive, I tried it and it works like a charm. Thank you very much!!!

**Comment by nicksergeant** [comment]
I'd actually love it if ScriptKit provided a way to do this. The downside of using the built-in macOS Keyboard -> Shortcuts -> App Shortcuts approach is that it is not backed up as part of iCloud, so changing machines or syncing between two machines isn't possible. I use Keyboard Maestro to do this now, but would be neat if ScriptKit could have "Emulate key command" or "Select menu item" actions for this.

**Reply by johnlindquist** [reply]
I use Karabiner extensively for managing keyboard shortcuts. I gave a talk my approach here: https://egghead.io/lessons/egghead-save-your-hands-and-save-your-time-rethinking-how-to-use-a-keyboard


"Select menu item" would be an interesting challenge. It would require wrapping some AppleScript just like in the "tileWindow" fn:

https://github.com/johnlindquist/kit/blob/7ee1cb127a6f8da8acba25cc8d76428a95eff9f5/src/lib/desktop.ts#L305

We'd have to make a comma/whatever separated list of menus to click 🤔


There's a "keystroke" fn if that's what you mean by "emulate".

https://github.com/johnlindquist/kit/blob/7ee1cb127a6f8da8acba25cc8d76428a95eff9f5/src/lib/keyboard.ts#L11

**Comment by nicksergeant** [comment]
Fascinating! Looks like I could use `keystroke` to accomplish this...

But last night I also ended up using the same AppleScript API in this way:

```
tell application "System Events"
	set appName to name of first application process whose frontmost is true
	if "Activity Monitor" is in appName then
		keystroke "f" using {option down, command down}
	end if
end tell
```

The problem that I have with this approach is that I can't "scope this rebinding to a specific application, otherwise do nothing and let the key trigger do its normal behavior".

In Keyboard Maestro I have this working by putting key remappings into groups that are only active in specific applications, like so:

## Grouping

![image](https://user-images.githubusercontent.com/27332/143768992-54e00ddc-b859-4855-993b-f11d48820615.png)

## Macro

![image](https://user-images.githubusercontent.com/27332/143769004-ce771fbb-be39-4e7f-bd46-abdd5413a74b.png)

This way I can say "cmd-f while in Activity Monitor triggers cmd-option-f" and "cmd-f while in Spotify triggers cmd-l", and cmd-f in any other application will be unaffected.

While I _love_ Keyboard Maestro and would be lost without it, I'd love a way to do this in a way that is scriptable perhaps in ScriptKit.

**Reply by nicksergeant** [reply]
Though, in my Activity Monitor macro here I'm selecting a menu item, but in my Spotify macro for the same keyboard shortcut, I'm manually triggering a keystroke:

![image](https://user-images.githubusercontent.com/27332/143769120-4b10a4d3-4d18-4275-af23-7137d887710f.png)

**Reply by nicksergeant** [reply]
(and sorry, forgot to reply directly to my "answer" above... first time using GH Discussions 😆)

**Reply by johnlindquist** [reply]
Here's a snippet that can get the frontmost app name using AppleScript:

```js
let app = await applescript(String.raw`
tell application "System Events"
  get name of first application process whose frontmost is true as text
end tell
`)
```

> Note: Next year we'll be moving all these system utils from AppleScript to Rust for speed and to support Windows. Also, I really, really dislike AppleScript :D

**Reply by nicksergeant** [reply]
@johnlindquist 👍 unfortunately it doesn't seem possible to trigger keystrokes conditionally by app while reverting to the default behavior of the initial keystroke as the following will just go into an infinite loop:

```applescript
tell application "System Events"
	set appName to name of first application process whose frontmost is true
	if "Activity Monitor" is in appName or "Mail" is in appName then
		keystroke "f" using {option down, command down}
	else
		keystroke "f" using {command down}
	end if
end tell
```

**Reply by johnlindquist** [reply]
I like the way you think 👍

This is something I do all the time in Karabiner. I would highly recommend investing time into learning Karabiner/goku.

(Feel free to ask me questions, I definitely want to make more lessons on it).

Here's my config. Search for "condi" and you'll see examples of conditional shortcuts based on a variety of factors (including frontmost app):

https://github.com/johnlindquist/dotfiles/blob/master/karabiner/karabiner.edn

**Reply by nicksergeant** [reply]
Yeah I may need to give Karabiner another shot again for these remappings. I remember managing unwieldy Karabiner configs before they made the move to Karabiner Elements 😅, after which I just switched everything to Keyboard Maestro as it can do those and more general purpose scripting. But, the KM UI leaves something to be desired and I worry about the lack of portability of my macros should KM go away.

**Reply by johnlindquist** [reply]
Yeah, using Goku has improved my Karabiner experience x1000

https://github.com/yqrashawn/GokuRakuJoudo

**Reply by nicksergeant** [reply]
Oh man this is awesome. The old style XML config of the original Karabiner (and what it was called before Karabiner, which is escaping me…) was _awful_ 😂.

Thanks for the tip!

**Reply by nicksergeant** [reply]
Ahh yes “KeyRemap4Macbook” 😆

**Reply by nicksergeant** [reply]
Following up here, Goku is indeed awesome, got all of my simple rebindings into Karabiner now: https://github.com/nicksergeant/dotfiles/blob/master/karabiner.edn

Thanks again for the tip!
